src/trogdord/docs/configuration.md
# Configuration

All of trogdord's configuration takes place in `/etc/trogdord/trogdord.ini`. If you installed trogdord to a custom prefix, then this path will be relative to that location.

## Default trogdord.ini

Every new installation of trogdord includes the following default configuration file:

```ini
[network]

; By default, trogdord runs on port 1040. This setting can be overridden
; by the TROGDORD_PORT environment variable.
; port=1040

; Socket option to allow the socket to be bound to an address that's already in
; use. If you don't enable this, you might have to wait a couple of minutes
; after stopping trogdord before you can start it again. On the flip side, not
; enabling it might in some circumstances be safer. For an explanation of what
; this is doing, see:
; https://stackoverflow.com/questions/3229860/what-is-the-meaning-of-so-reuseaddr-setsockopt-option-linux
;
; This setting can be overridden by the TROGDORD_REUSE_ADDRESS environment
; variable.
; reuse_address=true

; Send TCP keep-alives (unless you know what you're doing, you'll want this
; option to remain on.) This setting can be overridden by the
; TROGDORD_SEND_TCP_KEEPALIVE environment variable.
; send_keepalive=true

; Tells trogdord which ip addresses to listen for when accepting connections.
; 0.0.0.0 tells trogdord to accept connections from anywhere via IPv4, while
; :: accomplishes the same with IPv6. By default, trogdord will only listen
; on localhost (127.0.0.1 for IPv4 and ::1 for IPv6.) This setting can be
; overridden by the TROGDORD_LISTEN_IPS environment variable.
; listen=["127.0.0.1", "::1"]

[logging]

; Trogdord support 3 different logging methods: stdout (standard out), stderr
; (standard error), and file (in which case you will set the logto option to
; the filename where you'd like your logs to be appended.) By default, logto is
; set to stderr. This setting can be overridden by the TROGDORD_LOGTO
; environment variable.
; logto=stderr
; logto=stdout
; logto=var/log/trogdord.log

[input]

; Input listeners receive strings from external sources asynchronously and feed
; them to the input driver. Zero or more input listeners can run at the same
; time. If redis support is enabled at compile-time, the redis input listener
; can be started by adding "redis" (the quotes are required) to the array of
; listeners below. The redis input listener can be further configured in the
; [redis] section. By default, no input listeners are configured.
;
; This setting can be overridden by the TROGDORD_INPUT_LISTENERS environment
; variable.
; listeners=[]
; listeners=["redis"]

[output]

; This setting controls how messages in the output buffer are stored and
; retrieved. By default, the selected driver is set to "local", which manages
; messages locally inside trogdord. The other available driver is "redis", which
; sends output messages to a redis channel (if you're sending output messages
; to a websocket, this makes for quick and easy integration with servers like
; Socket.IO.)
;
; This setting can be overridden by the TROGDORD_OUTPUT_DRIVER environment
; variable.
; driver=local

; These settings will only be used if redis support was enabled at compile-time
; and if the redis output driver or input listener are used.
[redis]

; The hostname where the redis instance is running. Defaults to localhost.
; This setting can be overridden by the TROGDORD_REDIS_HOST environment
; variable.
; host=localhost

; The port the redis instance is listening on. This setting can be
; overridden by the TROGDORD_REDIS_PORT environment variable.
; port=6379

; If required, this should be uncommented and set to the username needed to
; authenticate with Redis. This setting can be overridden by the
; TROGDORD_REDIS_USERNAME environment variable.
; username=somebody

; If required, this should be uncommented and set to the password needed to
; authenticate with Redis. This setting can be overridden by the
; TROGDORD_REDIS_PASSWORD environment variable.
; password=mypassword

; The number of milliseconds we should wait when attempting to connect to Redis
; before timing out. This setting can be overridden by the
; TROGDORD_REDIS_CONNECTION_TIMEOUT environment variable.
; connection_timeout=5000

; If a connection to redis fails, trogdord will automatically try to connect
; again after this number of milliseconds. This setting can be overridden by
; the TROGDORD_REDIS_CONNECTION_RETRY_INTERVAL environment variable.
; connection_retry_interval=5000

; The redis channel we should publish output messages to. This setting can be
; overridden by the TROGDORD_REDIS_OUTPUT_CHANNEL environment variable.
; output_channel=trogdord:out

; The redis channel we should subscribe to in order to receive input. This
; setting can be overridden by the TROGDORD_REDIS_INPUT_CHANNEL environment
; variable.
; input_channel=trogdord:in

[resources]

; This is the default path where game definition files are stored. Relative
; paths are based on the root directory where trogdord was installed. This
; setting can be overridden by the TROGDORD_DEFINITIONS_PATH environment
; variable.
; definitions_path=share/trogdor

[state]

; Whether or not to enable the saving and loading of game and server state. This
; is off by default. This setting can be overridden by the TROGDORD_STATE_ENABLED
; environment variable.
; enabled=false

; This is the serialization format that should be used when saving a game's
; state. The two built-in formats (if enabled during compilation) are json
; and sqlite. This setting can be overridden by the TROGDORD_STATE_FORMAT
; environment variable.
; format=json
; format=sqlite

; This is the default path where game state is saved. Relative paths are based
; on the root directory where trogdord was installed. This setting can be
; overridden by the TROGDORD_STATE_PATH environment variable.
; save_path=var/trogdord/state

; This is the maximum number of dumps to store for each game. If set to 0, we
; keep all previous dumps. This setting can be overridden by the
; TROGDORD_STATE_MAX_DUMPS_PER_GAME environment variable.
; max_dumps_per_game=1

; If state is enabled, this determines whether or not we should automatically
; restore the server's last known state the next time it starts up. This is
; disabled by default. This setting can be overridden by the
; TROGDORD_STATE_AUTORESTORE_ENABLED environment variable.
; auto_restore=false

; If set to true, and if the state feature is enabled, the server's state will
; be dumped to disk on shutdown. This setting can be overridden by the
; TROGDORD_STATE_DUMP_SHUTDOWN_ENABLED environment variable.
; dump_on_shutdown=false

; If set to true, and if the state feature is enabled, the server will attempt
; to dump its state to disk before crashing due to uncaught exceptions or other
; unusual behavior. This will not be invoked for things like SIGSEGV, SIGFPE,
; etc. I tried, and it doesn't work. This setting can be overridden by the
; TROGDORD_STATE_CRASH_RECOVERY_ENABLED environment variable.
; crash_recovery=false

; Currently, extensions are only supported on operating systems that use libdl
; (this excludes Windows.)
[extensions]

; This is the default path where trogdord extensions are stored. Relative paths
; are based on the root directory where trogdord was installed. This setting
; can be overridden by the TROGDORD_EXTENSIONS_PATH environment variable.
; path=lib/trogdord

; Extensions to be loaded should be formatted as a JSON array of strings, and
; each extension should be referenced by its full filename, including
; extension. For example: ["ext1.so", 'ext2.so']
;
; You can find a buildable, loadable sample extension, which can serve as a
; template for your own projects, in sample/extensions/hello.
;
; This setting can be overridden by the TROGDORD_EXTENSIONS_LOAD environment
; variable.
; load=[]
```

To change an option from its default value, uncomment it by deleting the semi-colon, set it to the desired value, save the file, and restart trogdord.

Changelog.md

# Change Log

## [0.92.0] - In Development

### Changed

- Entities no longer have their own Lua states, which is wasteful. Instead, there is only one shared state per game.

## [0.91.0] - 2022-03-06

### Added

- trogdord can now authenticate with redis using an optional username and password
- trogdord can now also be configured via environment variables, which take precedence over trogdord.ini

### Changed

- The timer period (tick interval) can now be changed after the timer's started and can be configured via a game definition file
- trogdord now only listens on 127.0.0.1 or ::1 by default

### Fixed

- Added missing validation for ALLOCATE_RESOURCE AST operation

## [0.90.0] - 2022-01-31

### Added

- Added room connection descriptions ([Connection descriptions feature #72 ](https://github.com/crankycyclops/trogdor-pp/pull/72)), a more flexible way to generate room descriptions that can be updated dynamically when those connections change. This new feature shouldn't have any impact on older projects.
- Wrote two new Lua methods, Being:insertIntoInventory() and Being:removeFromInventory() to replace functionality previously served by the now removed method Thing:setLocation.
- Implemented Lua function Resource.new() to create new Resources from within Lua
- Implemented Lua methods Game::getEntity(), Game:start(), Game:stop(), and Game::inProgress()
- Added LuaState::getLuaVersion(), which returns the version of Lua the core library was built against
- Added LuaEntity::err() to allow writing to an Entity's error stream
- Started writing unit tests for the LuaState class
- Unit tests are now built and run against Lua 5.1, 5.2, and 5.3
- trogdord can now listen on IPv6 as well as IPv4, and can also be configured to listen for requests only from certain IP addresses

### Changed

- Reimplemented the items in a Being's inventory using std::weak_ptr. Any code that calls Being::getInventoryObjects() or Being::getInventoryObjectsByName() will have to be adjusted to take this change into account.
- Game::removeEntity() now enforces a more complete set of rules and no longer returns a boolean value (instead, it throws an exception if anything goes wrong.)
- Object::setOwner() is now protected and an Object can only be inserted into a Being's inventory by calling Being::insertIntoInventory().
- Thing::setLocation() is now protected and a Thing can only be inserted into a Place by calling Place::insertThing().
- Removed Lua method Thing:setLocation().
- Many formerly private members and methods in trogdor::LuaState are now protected to make unit testing easier

### Fixed

- It turns out that the Tangible and Resource types weren't being registered with Lua. This was uncovered by my new unit tests and has been fixed.
- Fixed a bug that affected the way the Lua Resource type was registered
- Did some minor refactoring and code cleanup

## [0.80.0] - 2021-07-15

### Added

- Added extension loading for custom functionality to trogdord on platforms that support libdl.

### Changed

- Split the build into multiple CMake files, making things more modular (I'm hoping to eventually spin each item out into its own repository)

## [0.70.0] - 2021-06-22

### Added

- Introduced the SQLite3 serialization format and improves trogdord's handling of dumped games.

## [0.60.0] - 2021-03-16

### Added

- Introduced state features, making it possible to dump and restore game state to disk.

### Changes

- Many changes that I didn't track at the time

### Fixes

- Many fixes that I also didn't track at the time

## [0.50.2] - Before 2021-03-16

This was the version of the engine right before I started tracking versions and their changes. If you're really curious, you can look through the very lengthy commit log for more history.


src/trogdord/config.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>

#include <trogdord/json.h>
#include <trogdor/utility.h>
#include <trogdor/filesystem.h>

#include <trogdord/config.h>
#include <trogdord/filesystem.h>

#ifndef TROGDORD_INI_PATH
	#define TROGDORD_INI_PATH ""
#endif


// Maps env variables to config options they override
const std::unordered_map<std::string, std::string> Config::ENV = {

	{"TROGDORD_PORT",                             CONFIG_KEY_PORT},
	{"TROGDORD_REUSE_ADDRESS",                    CONFIG_KEY_REUSE_ADDRESS},
	{"TROGDORD_SEND_TCP_KEEPALIVE",               CONFIG_KEY_SEND_TCP_KEEPALIVE},
	{"TROGDORD_LISTEN_IPS",                       CONFIG_KEY_LISTEN_IPS},
	{"TROGDORD_LOGTO",                            CONFIG_KEY_LOGTO},
	{"TROGDORD_INPUT_LISTENERS",                  CONFIG_KEY_INPUT_LISTENERS},
	{"TROGDORD_OUTPUT_DRIVER",                    CONFIG_KEY_OUTPUT_DRIVER},
	{"TROGDORD_REDIS_HOST",                       CONFIG_KEY_REDIS_HOST},
	{"TROGDORD_REDIS_USERNAME",                   CONFIG_KEY_REDIS_USERNAME},
	{"TROGDORD_REDIS_PASSWORD",                   CONFIG_KEY_REDIS_PASSWORD},
	{"TROGDORD_REDIS_PORT",                       CONFIG_KEY_REDIS_PORT},
	{"TROGDORD_REDIS_CONNECTION_TIMEOUT",         CONFIG_KEY_REDIS_CONNECTION_TIMEOUT},
	{"TROGDORD_REDIS_CONNECTION_RETRY_INTERVAL",  CONFIG_KEY_REDIS_CONNECTION_RETRY_INTERVAL},
	{"TROGDORD_REDIS_OUTPUT_CHANNEL",             CONFIG_KEY_REDIS_OUTPUT_CHANNEL},
	{"TROGDORD_REDIS_INPUT_CHANNEL",              CONFIG_KEY_REDIS_INPUT_CHANNEL},
	{"TROGDORD_DEFINITIONS_PATH",                 CONFIG_KEY_DEFINITIONS_PATH},
	{"TROGDORD_STATE_ENABLED",                    CONFIG_KEY_STATE_ENABLED},
	{"TROGDORD_STATE_AUTORESTORE_ENABLED",        CONFIG_KEY_STATE_AUTORESTORE_ENABLED},
	{"TROGDORD_STATE_DUMP_SHUTDOWN_ENABLED",      CONFIG_KEY_STATE_DUMP_SHUTDOWN_ENABLED},
	{"TROGDORD_STATE_CRASH_RECOVERY_ENABLED",     CONFIG_KEY_STATE_CRASH_RECOVERY_ENABLED},
	{"TROGDORD_STATE_FORMAT",                     CONFIG_KEY_STATE_FORMAT},
	{"TROGDORD_STATE_PATH",                       CONFIG_KEY_STATE_PATH},
	{"TROGDORD_STATE_MAX_DUMPS_PER_GAME",         CONFIG_KEY_STATE_MAX_DUMPS_PER_GAME},
	{"TROGDORD_EXTENSIONS_PATH",                  CONFIG_KEY_EXTENSIONS_PATH},
	{"TROGDORD_EXTENSIONS_LOAD",                  CONFIG_KEY_EXTENSIONS_LOAD}
};

// Default ini values
const std::unordered_map<std::string, std::string> Config::DEFAULTS = {

	{CONFIG_KEY_PORT,                             "1040"},
	{CONFIG_KEY_REUSE_ADDRESS,                    "true"},
	{CONFIG_KEY_SEND_TCP_KEEPALIVE,               "true"},
	{CONFIG_KEY_LISTEN_IPS,                       "[\"127.0.0.1\", \"::1\"]"},
	{CONFIG_KEY_LOGTO,                            "stderr"},
	{CONFIG_KEY_INPUT_LISTENERS,                  "[]"},
	{CONFIG_KEY_OUTPUT_DRIVER,                    "local"},
	{CONFIG_KEY_REDIS_HOST,                       "localhost"},
	{CONFIG_KEY_REDIS_USERNAME,                   ""},
	{CONFIG_KEY_REDIS_PASSWORD,                   ""},
	{CONFIG_KEY_REDIS_PORT,                       "6379"},
	{CONFIG_KEY_REDIS_CONNECTION_TIMEOUT,         "5000"},
	{CONFIG_KEY_REDIS_CONNECTION_RETRY_INTERVAL,  "5000"},
	{CONFIG_KEY_REDIS_OUTPUT_CHANNEL,             "trogdord:out"},
	{CONFIG_KEY_REDIS_INPUT_CHANNEL,              "trogdord:in"},
	{CONFIG_KEY_DEFINITIONS_PATH,                 "share/trogdor"},
	{CONFIG_KEY_STATE_ENABLED,                    "false"},
	{CONFIG_KEY_STATE_AUTORESTORE_ENABLED,        "false"},
	{CONFIG_KEY_STATE_DUMP_SHUTDOWN_ENABLED,      "false"},
	{CONFIG_KEY_STATE_CRASH_RECOVERY_ENABLED,     "false"},
	{CONFIG_KEY_STATE_FORMAT,                     "json"},
	{CONFIG_KEY_STATE_PATH,                       "var/trogdord/state"},
	{CONFIG_KEY_STATE_MAX_DUMPS_PER_GAME,         "0"},
	{CONFIG_KEY_EXTENSIONS_PATH,                  "lib/trogdord"},
	{CONFIG_KEY_EXTENSIONS_LOAD,                  "[]"}
};

// Any setting that has a value of true will be considered sensitive
// information and will be hidden from the outside world.
const std::unordered_map<std::string, bool> Config::hidden = {

	{CONFIG_KEY_PORT,                             true},
	{CONFIG_KEY_REUSE_ADDRESS,                    true},
	{CONFIG_KEY_SEND_TCP_KEEPALIVE,               true},
	{CONFIG_KEY_LISTEN_IPS,                       true},
	{CONFIG_KEY_LOGTO,                            true},
	{CONFIG_KEY_INPUT_LISTENERS,                  true},
	{CONFIG_KEY_OUTPUT_DRIVER,                    false},
	{CONFIG_KEY_REDIS_HOST,                       true},
	{CONFIG_KEY_REDIS_USERNAME,                   true},
	{CONFIG_KEY_REDIS_PASSWORD,                   true},
	{CONFIG_KEY_REDIS_PORT,                       true},
	{CONFIG_KEY_REDIS_CONNECTION_TIMEOUT,         true},
	{CONFIG_KEY_REDIS_CONNECTION_RETRY_INTERVAL,  true},
	{CONFIG_KEY_REDIS_OUTPUT_CHANNEL,             false},
	{CONFIG_KEY_REDIS_INPUT_CHANNEL,              false},
	{CONFIG_KEY_DEFINITIONS_PATH,                 true},
	{CONFIG_KEY_STATE_ENABLED,                    false},
	{CONFIG_KEY_STATE_AUTORESTORE_ENABLED,        true},
	{CONFIG_KEY_STATE_DUMP_SHUTDOWN_ENABLED,      true},
	{CONFIG_KEY_STATE_CRASH_RECOVERY_ENABLED,     true},
	{CONFIG_KEY_STATE_FORMAT,                     true},
	{CONFIG_KEY_STATE_PATH,                       true},
	{CONFIG_KEY_STATE_MAX_DUMPS_PER_GAME,         false},
	{CONFIG_KEY_EXTENSIONS_PATH,                  true},
	{CONFIG_KEY_EXTENSIONS_LOAD,                  false}
};

const std::unordered_map<std::string, const std::type_info *> Config::types = {

	{CONFIG_KEY_PORT,                             &typeid(int)},
	{CONFIG_KEY_REUSE_ADDRESS,                    &typeid(bool)},
	{CONFIG_KEY_SEND_TCP_KEEPALIVE,               &typeid(bool)},
	{CONFIG_KEY_LISTEN_IPS,                       &typeid(std::string)},
	{CONFIG_KEY_LOGTO,                            &typeid(std::string)},
	{CONFIG_KEY_INPUT_LISTENERS,                  &typeid(std::string)},
	{CONFIG_KEY_OUTPUT_DRIVER,                    &typeid(std::string)},
	{CONFIG_KEY_REDIS_HOST,                       &typeid(std::string)},
	{CONFIG_KEY_REDIS_USERNAME,                   &typeid(std::string)},
	{CONFIG_KEY_REDIS_PASSWORD,                   &typeid(std::string)},
	{CONFIG_KEY_REDIS_PORT,                       &typeid(int)},
	{CONFIG_KEY_REDIS_CONNECTION_TIMEOUT,         &typeid(int)},
	{CONFIG_KEY_REDIS_CONNECTION_RETRY_INTERVAL,  &typeid(int)},
	{CONFIG_KEY_REDIS_OUTPUT_CHANNEL,             &typeid(std::string)},
	{CONFIG_KEY_REDIS_INPUT_CHANNEL,              &typeid(std::string)},
	{CONFIG_KEY_DEFINITIONS_PATH,                 &typeid(std::string)},
	{CONFIG_KEY_STATE_ENABLED,                    &typeid(bool)},
	{CONFIG_KEY_STATE_AUTORESTORE_ENABLED,        &typeid(bool)},
	{CONFIG_KEY_STATE_DUMP_SHUTDOWN_ENABLED,      &typeid(bool)},
	{CONFIG_KEY_STATE_CRASH_RECOVERY_ENABLED,     &typeid(bool)},
	{CONFIG_KEY_STATE_FORMAT,                     &typeid(std::string)},
	{CONFIG_KEY_STATE_PATH,                       &typeid(std::string)},
	{CONFIG_KEY_STATE_MAX_DUMPS_PER_GAME,         &typeid(int)},
	{CONFIG_KEY_EXTENSIONS_PATH,                  &typeid(std::string)},
	{CONFIG_KEY_EXTENSIONS_LOAD,                  &typeid(std::string)}
};

// Singleton instance of Config
std::unique_ptr<Config> Config::instance = nullptr;

/*****************************************************************************/

Config::Config(std::string newIniPath) noexcept {

	// Load the ini file's values
	load(newIniPath);
}

/*****************************************************************************/

std::unique_ptr<Config> &Config::get() noexcept {

	if (!instance) {
		instance = std::unique_ptr<Config>(new Config(TROGDORD_INI_PATH));
	}

	return instance;
}

/*****************************************************************************/

std::string Config::relativeToAbsolutePath(std::string relative) {

	relative = trogdor::trim(relative);

	if (0 != relative.compare("")) {
		relative = Filesystem::getAbsolutePath(relative);
	}

	return relative;
}

/*****************************************************************************/

std::vector<std::string> Config::getExtensions() {

	rapidjson::Document extensionsArr;

	std::string extensionsStr = Config::get()->getString(
		Config::CONFIG_KEY_EXTENSIONS_LOAD
	);

	extensionsArr.Parse(extensionsStr.c_str());

	if (extensionsArr.HasParseError()) {
		throw ServerException(
			std::string("invalid trogdord.ini value for ") +
			Config::CONFIG_KEY_EXTENSIONS_LOAD
		);
	}

	else if (rapidjson::kArrayType != extensionsArr.GetType()) {
		throw ServerException(
			std::string("trogdord.ini value for ") +
			Config::CONFIG_KEY_EXTENSIONS_LOAD +
			" must be a JSON array of strings."
		);
	}

	std::vector<std::string> extensions;

	for (auto i = extensionsArr.Begin(); i != extensionsArr.End(); i++) {
		extensions.push_back(i->GetString());
	}

	return extensions;
}

/*****************************************************************************/

void Config::initErrorLogger() noexcept {

	logFileStream = nullptr;
	std::string logto = trogdor::strToLower(getString(CONFIG_KEY_LOGTO));

	if (0 == logto.compare("stderr")) {
		errStream = std::make_unique<ServerErr>(&std::cerr);
	}

	else if (0 == logto.compare("stdout")) {
		errStream = std::make_unique<ServerErr>(&std::cout);
	}

	else {

		logto = Filesystem::getAbsolutePath(logto);

		try {
			logFileStream = std::make_unique<std::ofstream>(logto, std::ofstream::app);
			errStream = std::make_unique<ServerErr>(logFileStream.get());
		}

		// If for whatever reason we failed to open the file for writing
		// (maybe the process doesn't have the necessary permissions?), fall
		// back to std::cerr.
		catch (const std::exception &e) {

			std::cerr << e.what() << std::endl;
			std::cerr << "WARNING: failed to open " << logto << " for writing. Falling back to stderr." << std::endl;

			logFileStream = nullptr;
			errStream = std::make_unique<ServerErr>(&std::cerr);
		}
	}
}

/*****************************************************************************/

void Config::load(std::string newIniPath) noexcept {

	iniPath = trogdor::trim(newIniPath);
	ini.clear();

	// Start by populate the ini object with its default values
	for (auto &defaultVal: DEFAULTS) {
		ini[defaultVal.first] = defaultVal.second;
	}

	// Next, override the defaults with any options in the ini file
	if (
		0 != iniPath.compare("") &&
		STD_FILESYSTEM::exists(iniPath) &&
		!STD_FILESYSTEM::is_directory(iniPath)
	) {
		int status = ini_parse(iniPath.c_str(), [](
			void *data,
			const char *section,
			const char *name,
			const char *value
		) -> int {

			std::string key = section ? std::string(section) + '.' + name : name;

			static_cast<Config *>(data)->ini[key] = value;
			return 1;

		}, this);

		if (-1 == status) {
			std::cerr << "Error: " << "Couldn't open " << iniPath
				<< " for reading." << std::endl;
			exit(EXIT_INI_ERROR);
		}

		else if (-2 == status) {
			std::cerr << "Error: " << "A memory allocation error occured when attempting to parse "
				<< iniPath << '.' << std::endl;
			exit(EXIT_INI_ERROR);
		}

		else if (0 != status) {
			std::cerr << "Error: " << "An unknown error occured when attempting to parse "
				<< iniPath << '.' << std::endl;
		}
	}

	// Finally, override any settings that have been configured in environment
	// variables (these take precedence over trogdord.ini.)
	for (auto &envVar: ENV) {
		if (const char *setting = std::getenv(envVar.first.c_str()); setting) {
			ini[envVar.second] = setting;
		}
	}

	// Setup global error logger
	initErrorLogger();
}
